app/notifiers/notifier/base.rb

```diff
@@ -0,0 +1,71 @@
+# The base class of notifiers. This is meant to be called by the Notifications Framework
+#
+# @api notifications
+class Notifier::Base
+  class ActivityWrapper < SimpleDelegator
+    # Send notifications according to input type and recipient
+    #
+    # @param [Object] recipient The recipient of the notification
+    # @param [Symbol] type The type of notification
+    def notify(recipient, type)
+      super(recipient, type).tap do |notification|
+        case recipient
+        when Course
+          emails = @notifier.send(:notify_course, notification)
+          @pending_emails.push(emails) unless emails.nil?
+        when User
+          email = @notifier.send(:notify_user, notification)
+          @pending_emails.push(email) unless email.nil?
+        end
+      end
+
+      self
+    end
+
+    # Save activity and send out pending emails
+    def save(*)
+      super.tap do |result|
+        @pending_emails.each(&:deliver_later) if result
+      end
+    end
+
+    private
+
+    def initialize(notifier, activity) #:nodoc:
+      super(activity)
+      @notifier = notifier
+      @pending_emails = []
+    end
+  end
+
+  class << self
+    # This is not meant to be called by user directly
+    def method_missing(symbol, *args, &block)
+      new.send(symbol, *args, &block)
+    end
+  end
+
+  private
+
+  def create_activity(options) #:nodoc:
+    ActivityWrapper.new(self, Activity.new(options.merge(notifier_type: self.class.name)))
+  end
+
+  def notify_course(notification) #:nodoc:
+    case notification.notification_type
+    when 'email'
```

Comment:
I think should use `if` here, there is only one case.

And document these 3 methods, cause they are important.